modules/biobank/php/biobank.class.inc

```diff
@@ -0,0 +1,120 @@
+<?php declare(strict_types=1);
+/**
+ * This class features the code that enables access to the LORIS Biobank Module.
+ *
+ * PHP Version 7.2
+ *
+ * @category   Page
+ * @package    Loris
+ * @subpackage Biobank
+ * @author     Henri Rabalais <henri.rabalais@mcin.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.text GPLv3
+ * @link       http://www.github.com/aces/Loris/
+ */
+namespace LORIS\biobank;
+use \Psr\Http\Message\ServerRequestInterface;
+use \Psr\Http\Message\ResponseInterface;
+
+/**
+ * This class features the code that enables access to the LORIS Biobank Module.
+ *
+ * PHP Version 7.2
+ *
+ * @category   Page
+ * @package    Loris
+ * @subpackage Biobank
+ * @author     Henri Rabalais <henri.rabalais@mcin.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.text GPLv3
+ * @link       http://www.github.com/aces/Loris/
+ */
+class Biobank extends \NDB_Menu_Filter
+{
+    /**
+     * Determine whether the user has permission to view this page.
+     *
+     * @param User $user to be checked.
+     *
+     * @return bool whether the user has access.
+     */
+    function _hasAccess(\User $user) : bool
+    {
+        return $user->hasPermission('biobank_access');
+    }
+
+    /**
+     * Set up the variables required by for construction of page
+     *
+     * @return void
+     */
+    function _setupVariables()
+    {
+        $this->skipTemplate = true;
+        $this->AjaxModule   = true;
+    }
+
+    // TODO: Revisit file uploading mechanism. This should likely be done
+    // somewhere else using different code. Perhaps John S. can provide some
+    // clarity.
+    /**
+     * Allows files to be downloaded
+     *
+     * @param file $file File to be downloaded
+     * @param user $user User to be checked for permission
+     *
+     * @return ResponseInterface the outgoing PSR7 response.
+     */
+    function downloadFile($file, $user) : ResponseInterface
+    {
+        if (!$user->hasPermission('biobank_specimen_view')) {
+            throw new \Forbidden('Permission to download file is denied');
+        }
+
+        $fileName = basename($file);
+        $config   = \NDB_Config::singleton();
+        $path     = $config->getSetting('mediaPath');
+        $filePath = $path . $fileName;
+
+        if (!file_exist($filePath)) {
+            error_log('ERRO: File'.$filePath.' does not exist');
+            throw new \BadRequest('File was not found');
+        }
+
+        return (new \LORIS\Http\Response())
+            ->withHeader('Content-Type', 'image/jpeg')
+            ->withHeaer('Content-Length', (string) filesize($file))
+            ->withBody(new \LORIS\Http\StringStream($filePath));
```

Comment:
I haven't tested this recently, and I'm not really sure this function works at all. I will likely move it to another class explicitly for file downloading. I was hoping there was already a file download class so I wouldn't have to make one ahah.

Reply:
Yeah, I'm pretty sure this won't work..

Reply:
I can't seem to figure out how to download files using PSR7 responses. Do you have any documentation references for this @driusan?

Reply:
You need a Psr\Http\Message\StreamInterface implementation which implements the interface as a wrapper around fread/fwrite/etc. Since the interface is basically just an object oriented interface version of fread/fwrite/etc calls in PHP, you can either write one yourself pretty trivially, or I suspect Zend (the PSR7 implementation we're currently using) already includes one that you could write a small wrapper around.

Reply:
@HenriRabalais I'll be resuming work on the File Download class when the Upload class is finalized and that'll take care of what @driusan is explaining above. I'll come back to you when that process is done and we can update this portion of the code.

Reply:
Sounds good. I will remove this code and resolve for now since it is not functional and not being called anywhere.